syntax="proto3";

package gate;

message TestReq {
    int32   service_id          = 1;
}

message TestRsp {
    bytes   service_name        = 1;
}

message WakeHeartbeatReq {
    uint64  server_time         = 1;
}

message WakeHeartbeatRsp {
    uint64  client_time         = 1;
    uint64  key                 = 2;    // server_time xor conn_id
}

message GetConnIdReq {
    uint64  client_time         = 1;    // 客户端的时间
}

message GetConnIdRsp {
    uint64  server_time         = 1;    // 服务端的时间
    uint32  conn_create_time    = 2;    // 客户端连接创建时间
    uint64  conn_id             = 3;    // 客户端的连接id
}

message CloseConnIdReq {
    uint64  conn_id             = 1;
}

message CloseConnIdRsp {

}

message GateMsg {                                   // 奇数是请求，偶数是响应
    oneof choice {
        TestReq         test_req            = 1;    // 请求
        TestRsp         test_rsp            = 2;    // 响应

        WakeHeartbeatReq    wake_heartbeat_req  = 3;    // 唤醒客户端心跳
        WakeHeartbeatRsp    wake_heartbeat_rsp  = 4;

        GetConnIdReq    get_conn_id_req     = 5;    // 获取连接id
        GetConnIdRsp    get_conn_id_rsp     = 6;

        CloseConnIdReq  close_conn_id_req   = 7;    // 关闭连接id
        CloseConnIdRsp  close_conn_id_rsp   = 8;
    }
}